### PR TITLE
build: T3664: add support for building non-ISO flavors

### PR DESCRIPTION
**Pull request by dmbaturin**
## Change Summary

Add support for building flavors other than ISO images directly from vyos-build.

The diff is, unfortunately, unavoidably messy-looking due to a large change, sorry about that.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes)
- [ ] Migration from an old Vyatta component to vyos-1x, please link to related PR inside obsoleted component
- [ ] Other (please describe):

## Component(s) name

Build scripts.

## Proposed changes

The following new fields are available in flavor files:

```
# Image format: can be "raw" or anything supported by qemu-img
image_format = ["raw"]

# Post-build hook script for executing special procedures (such as OVA building signing).
# It gets the raw image path relative to build/ as its argument.
post_build_hook = "/path/to/script"

# Boot settings for GRUB configuration
[build_settings]
  timeout = 5
  console_type = 'tty',
  console_num = 0
  console_speed = 115200
  bootmode = 'normal'
```

The build script supports the following new options:

* `--disk-size` — disk size in gigabytes (ignored for ISO)
* `--reuse-iso` — path to an ISO image (relative to `build/`), if given, the script will use it instead of building it from scratch.

The script uses the following new environment variables:

* ` VYOS_TEMPLATE_DIR` — normally should point to the template dir of the `vyos-1x` submodule (the variable is passwd to `vyos.template`). This variable is currently required for the script to work, we may want to find a way to add a hardcoded default to it so simplify script calls.
* `VYOS_BUILD_FLAVORS_DIR` — optional variable to specify a separate dir with flavors (e.g., the new vyos-community-flavors repository),

## How to test

For example, put the following in `~/flavors/raw.toml`:

```
# Generic (aka "universal") raw image
# Like the generic ISO, but raw ;)

image_format = "raw"

# Include these packages in the image regardless of the architecture
packages = [
  # QEMU and Xen guest tools exist for multiple architectures
  "qemu-guest-agent",
  "vyos-xe-guest-utilities",
]

[architectures.amd64]
  # Hyper-V and VMware guest tools are x86-only
  packages = ["hyperv-daemons", "vyos-1x-vmware"]

```

You can download an nightly image to speed up things.

Then run something like this:

```
sudo VYOS_BUILD_FLAVORS_DIR=/home/dmbaturin/flavors/  VYOS_TEMPLATE_DIR=/home/dmbaturin/vyos-build/vyos-1x/data/templates/ ./build-vyos-image --reuse-iso vyos-1.5-rolling-202404040019-amd64.iso raw
```

## Checklist:

- [x] I have read the [**CONTRIBUTING**](https://github.com/vyos/vyos-1x/blob/current/CONTRIBUTING.md) document
- [x] I have linked this PR to one or more Phabricator Task(s)
- [x] My commit headlines contain a valid Task id
- [x] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly

### CODE REVIEW

File: scripts/image-build/build-vyos-image [hunk 1]
```diff
@@ -157,16 +179,19 @@ if __name__ == "__main__":
         else:
             parser.add_argument('--' + k, type=str, help=help_string, default=default_value_thunk())
```
Comment (sever-sever) [inline scripts/image-build/build-vyos-image:202]: `str` object is not callable,  `default_value_thunk()`
should be `default_value_thunk`?

File: scripts/image-build/build-vyos-image [hunk 2]
```diff
@@ -268,277 +305,301 @@ if __name__ == "__main__":
...
+                build_git += "-dirty"
+
+            # Retrieve git branch name or current tag
+            # Building a tagged release might leave us checking out a git tag that is not the tip of a named branch (detached HEAD)
+            # Check if the current HEAD is associated with a tag and use its name instead of an unavailable branch name.
+            git_branch = next((tag.name for tag in repo.tags if tag.commit == repo.head.commit), None)
+            if git_branch is None:
+                git_branch = repo.active_branch.name
+        except Exception as e:
+          exit(f'Could not retrieve information from git: {repr(e)}')
```
Comment (sever-sever) [inline scripts/image-build/build-vyos-image]: There are 2 whitespaces, expected 4

File: scripts/image-build/build-vyos-image [hunk 3]
```diff
@@ -268,277 +305,301 @@ if __name__ == "__main__":
...
+          build_git = ""
```
Comment (sever-sever) [inline scripts/image-build/build-vyos-image]: This code is probably unreachable due to the above exit

**Review by sever-sever — CHANGES_REQUESTED** [review]

**Comment by sever-sever** [comment]
Help should check if `vyos-1x` exists
```
vyos_bld@83a9e41894e5:/vyos/tmp/vyos-build$ sudo ./build-vyos-image --help
E: vyos-1x subdirectory does not exist, did you initialize submodules?
Traceback (most recent call last):
  File "/vyos/tmp/vyos-build/./build-vyos-image", line 56, in <module>
    import raw_image
  File "/vyos/tmp/vyos-build/scripts/image-build/raw_image.py", line 23, in <module>
    import vyos.utils.process
ModuleNotFoundError: No module named 'vyos'
vyos_bld@83a9e41894e5:/vyos/tmp/vyos-build$ 
```

Dependency
```
vyos_bld@83a9e41894e5:/vyos/tmp/vyos-build$ git clone https://github.com/vyos/vyos-1x

vyos_bld@83a9e41894e5:/vyos/tmp/vyos-build$ sudo ./build-vyos-image --help
I: Checking if packages required for VyOS image build are installed
E: E: There are missing system dependencies!
E: Missing packages: gdisk dosfstools
vyos_bld@83a9e41894e5:/vyos/tmp/vyos-build$ 

```

**Comment by sever-sever** [comment]
Unexpected flavor `README`
```
vyos_bld@83a9e41894e5:/vyos/tmp/vyos-build$ sudo ./build-vyos-image --build-by viacheslav@host.local --reuse-iso=./vyos-1.5-rolling-202404130016-amd64.iso 
I: Checking if packages required for VyOS image build are installed
I: using build flavors directory data/build-flavors
E: Build flavor is not specified!
E: For example, to build the generic ISO, run ./build-vyos-image iso
Available build flavors:

edgecore
dell-vep1400
README
xcpng
iso
dell-vep4600
aws-iso
azure-iso

```

**Comment by sever-sever** [comment]
It does nothing:
```
vyos_bld@83a9e41894e5:/vyos/tmp/vyos-build$ cat ../flavors/raw.toml 
# Generic (aka "universal") ISO image

image_format = "iso"

# Include these packages in the image regardless of the architecture
packages = [
  # QEMU and Xen guest tools exist for multiple architectures
  "qemu-guest-agent",
  "vyos-xe-guest-utilities",
]
vyos_bld@83a9e41894e5:/vyos/tmp/vyos-build$ 


vyos_bld@83a9e41894e5:/vyos/tmp/vyos-build$ sudo VYOS_BUILD_FLAVORS_DIR=../flavors/ ./build-vyos-image --build-by viacheslav@host.local --reuse-iso vyos-1.5-rolling-202404130016-amd64.iso raw
I: Checking if packages required for VyOS image build are installed
I: using build flavors directory ../flavors/
vyos_bld@83a9e41894e5:/vyos/tmp/vyos-build$ 

```

**Comment by c-po** [comment]
`gdisk dosfstools` should be added to the Docker container as build dependencies for vyos-build https://github.com/vyos/vyos-build/blob/current/docker/Dockerfile#L92-L93

**Comment by c-po** [comment]
> It does nothing:
> 
> ```
> vyos_bld@83a9e41894e5:/vyos/tmp/vyos-build$ cat ../flavors/raw.toml 
> # Generic (aka "universal") ISO image
> 
> image_format = "iso"
> 
> # Include these packages in the image regardless of the architecture
> packages = [
>   # QEMU and Xen guest tools exist for multiple architectures
>   "qemu-guest-agent",
>   "vyos-xe-guest-utilities",
> ]
> vyos_bld@83a9e41894e5:/vyos/tmp/vyos-build$ 
> 
> 
> vyos_bld@83a9e41894e5:/vyos/tmp/vyos-build$ sudo VYOS_BUILD_FLAVORS_DIR=../flavors/ ./build-vyos-image --build-by viacheslav@host.local --reuse-iso vyos-1.5-rolling-202404130016-amd64.iso raw
> I: Checking if packages required for VyOS image build are installed
> I: using build flavors directory ../flavors/
> vyos_bld@83a9e41894e5:/vyos/tmp/vyos-build$ 
> ```

I can second this:

```
(16:08) vyos_bld 09c0e4e65e29:/vyos/vyos-build [current] # sudo VYOS_BUILD_FLAVORS_DIR=/vyos/vyos-build VYOS_TEMPLATE_DIR=/vyos/vyos-1x/data/templates/ ./build-vyos-image --reuse-iso vyos-999.202404131522-amd64.iso raw
I: Checking if packages required for VyOS image build are installed
I: using build flavors directory /vyos/vyos-build
```

**Comment by dmbaturin** [comment]
>It does nothing

With `image_format = "iso"` and `--reuse-iso`, that's trying to build an ISO from an ISO. Since the ISO is already an ISO, the script doesn't have anything to do to produce an ISO.

**Review by sever-sever — APPROVED** [review]

File: scripts/image-build/build-vyos-image (hunk 1 above)
Comment (dmbaturin) [reply scripts/image-build/build-vyos-image:202]: The `default_value_thunk` is a function. A real example is `get_default_build_by` referenced in the `options` dict. Apparently PyCharm's parser isn't smart enough to understand that functions can be passed as arguments in Python.

File: .gitmodules [hunk 4]
```diff
@@ -0,0 +1,3 @@
+[submodule "vyos-1x"]
+	path = vyos-1x
```
Comment (c-po) [inline .gitmodules:2]: Do we really wan't to have vyos-1x submodule in the toplevel directory? A better place most likely would be under packages.

Comment (c-po) [reply .gitmodules:2]: Or would it be an option to clone the repo during build into the background to a well known location? e.g. under build?

Comment (dmbaturin) [reply .gitmodules:2]: I'll move the submodule to `packages/` and add a `git submodule update --init` in the background to free the user from having to clone it manually.

**Review by c-po — APPROVED** [review]

**Pull request merged**